### ISSUE TITLE
How do we deal with stateless components?

### ISSUE THREAD

**Issue by oliverwoodings** [QubitProducts/react-test-tree#49]
React 0.14 added stateless components. This is fine and dandy until you realise that stateless components behave very differently to regular components. This presents us with a few issues; some are easily solvable, some are not.

#### 1. Stateless components have no 'internal instance'
This is more of an internal concern. We use the internal instances of elements to access props, stores etc. Stateless components don't have an internal instance since they are literally just a pure function. We can solve this by wrapping stateless components in a regular component at render time:
```jsx
function wrap(statelessComponent) {
  return React.createClass({
    render: function() {
      return statelessComponent(this.props);
    }
  });
}
```

#### 2. Stateless components cannot contain refs
<img width="602" alt="screen shot 2015-10-29 at 09 20 30" src="https://cloud.githubusercontent.com/assets/621323/10814505/55e99480-7e1e-11e5-9fcc-ce798fc58bb1.png">
What this means for us:
* A test-tree node of a stateless component will be a basic node (no `get`, `getIn` etc)
* Having a stateless component in the middle of the render tree will create a void in the ref chain, meaning you cannot access lower-down components
Unfortunately there is really no way around this if we want to continue using `ref`s as our method of building the tree.

#### 3. Stateless components as children cannot have refs put on them
This is a biggie. This does not work:
```jsx
let Stateless = (props) => (
  <div />
);
let Wrapper = React.createClass({
  render() {
    return <Stateless ref="stateeee" />
  }
});
```
<img width="1198" alt="screen shot 2015-10-29 at 11 23 43" src="https://cloud.githubusercontent.com/assets/621323/10817088/8d41d1a2-7e2f-11e5-8892-325cd2eee9c9.png">
This means that users can't even add refs to stateless components. The other two points can be dealt with, however this issue means that react-test-tree literally cannot address stateless components in a tree.

### What this means for end users
Stateless components cannot be tested as children of other components since they cannot have refs put on them. They can only be tested if they are the direct input to the `testTree` function like so:
```
let Stateless = (props) => (
  <div />
);
var tree = testTree(<Stateless />);
```
Even then, the actual operations you could perform on them will be limited to what you could do to a basic element such as a div.

### The only possible solution...
The only way to solve all these issues completely is to switch from using refs to regular props. So instead of doing:
```jsx
let Blah = React.createClass({
  render() {
    return <div ref="foo" refCollection="bar" />
  }
});
```
you would have:

```jsx
let Blah = React.createClass({
  render() {
    return <div rtt="foo" rttCollection="bar" />
  }
});
```
Doing this switch would allow us to have complete control, giving us the ability to support refs and refCollections inside and outside stateless components.

@jhollingworth @Olical @joeybaker @goncalvesjoao @wbecker would love your thoughts on this

**Comment by oliverwoodings** [comment]
Ok this is worse than I thought - adding point number 3

**Comment by goncalvesjoao** [comment]
Yeah I felt this pain too when I realized that stateless components don't give rise to instances.

According to this: https://github.com/facebook/react/issues/4936

They are supposed to be that way.

And I thought: "Crap. Now I can't access Parent -> Child (Stateless) -> GrandChild component."
and then it hit me... hummm maybe that's the point, I guess unit tests shouldn't be doing this kind of expectations.

Each component should be tested isolated from all the others and therefore I probably shouldn't be writing a Parent test expectation based on its GrandChild.

"Ok, I guess there's actually no need for refs" I thought.

But then after I started testing a stateless component I realized that without refs in it, my unit-test would be coupled with the component's dom structure and if it changed my tests would break...

"That's not cool at all, how do we tests stateless components without refs?" So I ended up finding this issue: https://github.com/facebook/react/issues/4972

Which led me, to the conclusion that: stateless components are still fringe and I probably should be avoiding them for now...

**Comment by goncalvesjoao** [comment]
@oliverwoodings "rtt" and "rttCollection" would do the trick for me, but after reading https://github.com/facebook/react/issues/4936 something tells me that we are missing the point by doing that kind of workaround.

I can't find a good solution for testing stateless components without refs.... and this me making me very confused. I guess I need to study this stateless components stuff, some more.

**Comment by oliverwoodings** [comment]
@goncalvesjoao yeah this is exactly how my thought process went as well :p I understand their reasoning behind wanting to not have refs on stateless components when that code is running in a production environment. A stateless component should be opaque for everything _except_ when you are unit testing that specific component. Right now it seems like the component is opaque even when you want to test it's own functionality.

I think using `rtt` and `rttCollection` doesn't completely miss the point, since it only affects the testing environment. It's not as if test-tree is used to reach deep into rendered trees to call custom methods; it is simply used to make assertions and easily simulate user input.

You're probably right that stateless components should be treated as fringe right now. I mean, how on earth are you meant to test clicking the button like this:
```jsx
let Form = (props) => (
  <div>
    <button onClick={props.onSubmit} />
  </div>
);
```
The only way would be to scry the component out of the DOM, then trigger a native DOM click event. Ew. It seems like react needs a separate ptop to `ref` that is purely for testing. This was how I came up with the idea for `rtt` and `rttCollection`. We actually already use this technique for `refCollection` (since it is not native to react like `ref`).

**Comment by goncalvesjoao** [comment]
"I understand their reasoning behind wanting to not have refs on stateless components when that code is running in a production environment. A stateless component should be opaque for everything except when you are unit testing that specific component" nice, that's what I was missing.

Production stateless components should not rely on refs but that doesn't not mean that on a testing environment, one should be stripped of tools to achieve proper testing.

I agree, the onClick example is so simple and yet so ugly to test.

Personally I don't use refs for anything but tests, changing ref to rtt for me is straightforward since production code does not rely on refs, but I'd like to hear from other people their thoughts on this.

+ 1 on your proposition to use rtt and rttCollection.

**Comment by aguestuser** [comment]
Like @goncalvesjoao I use refs almost exclusively for testing. `rtt` and `rttCollection` would suit me fine. Question (which likely betrays a lack of familiarity with the implementation details at play): if the only purpose of these props is for testing, would it be possible to set some sort of transpile-time flag for webpack, uglify, etc... to strip them from the codebase for production use?

**Comment by joeybaker** [comment]
After reading through https://github.com/facebook/react/issues/4936 (FYI: https://github.com/facebook/react/issues/4972 is the issue to subscribe to), it seems like this whole problem is an oversight that React would like to solve. That said, there's no promise of a fix in the near future if at all.

The `rtt` and `rttCollection` idea seems fine, but I echo @aguestuser – it would be great to have a way to strip these props for production, maybe via a babel transform?

***

An aside:  I've run into a problem a lot like this while trying to test inside a redux `Provider` in react 0.13.

```js
testTree(<Provider store={store}>{ () => <MyComponent ref="hi" />}</Provider>) 
```

The above has a similar problem to stateless components – you can't access the `'hi'` ref with test tree.

**Comment by oliverwoodings** [comment]
Just had a good chat with @jhollingworth and @davidpadbury here at Qubit. We're going to go with the proposed solution, but will use `testRef` and `testRefCollection` since they are a bit more declarative. Babel transform also sounds like a great idea - we will look into it.

@joeybaker I believe that redux problem is slightly different - would you mind raising a separate issue? Thanks :)

**Comment by joeybaker** [comment]
:+1: Cool.

@oliverwoodings issue created: #50

**Comment by oliverwoodings** [comment]
@goncalvesjoao @aguestuser @joeybaker @jhollingworth do you think we should change `innerComponentRef`(the higher-order-component stuff) to use the new `testRef` prop? There are a few arguments either way:

#### Pros
* Less confusing - we wouldn't be mixing `testRef` and `ref`
* Will work with higher-order stateless components

#### Cons
* Quite a lot of 3rd party libraries already have `ref`s on their higher order components. Switching to using our bespoke `testRef` prop will mean those libraries won't be automatically compatible with react-test-tree.


I guess we could have both? We can keep `innerComponentRef` and also have `innerTestRef`?

**Comment by oliverwoodings** [comment]
I have released react-test-tree@1.0.0-rc4 with the new `testRef` and `testRefCollection` syntax. Would appreciate it if anyone could give it a try!

**Comment by jimfb** [comment]
A lot of the conversation seems to have been happening here, so I thought I'd jump in.

@goncalvesjoao said:
  > Which led me to the conclusion that: stateless components are still fringe and I probably should be avoiding them for now...

We (React team) does not see stateless components as fringe.  In fact, the community seems to be drifting more toward "If it is possible for a component to be a stateless function, it should be a stateless function".  Stateless functions are simpler, less boiler plate, easier to understand, more optimizable, etc.

PS: If anything, `ref`s should be seen as fringe and avoided; they are intended to be an escape hatch for getting access to the underlying component in order to do things like perform imperative operations (ideally you avoid all imperative operations) or access other APIs that we do not yet expose in React.

Having said all that, we're still working on a better story for testing.

---

@oliverwoodings said:
 > I mean, how on earth are you meant to test clicking the button like this...

`button`s are not stateless (yet), so you should be able to use callback refs to save a reference to the button and do whatever you want to it.  Is there a reason that does not work?  You would need to decide where you want to save the ref, but that's a very solvable problem.  FWIW, string refs (eg. `component.refs.myrefname`) are mostly legacy at this point anyway.  String refs are not deprecated yet, but the plan is to phase them out eventually.

---
@oliverwoodings said:
 > I understand their reasoning behind wanting to not have refs on stateless components when that code is running in a production environment. A stateless component should be opaque for everything except when you are unit testing that specific component.

We have an internal requirement/philosophy that components behave identically (or as close as possible) in dev and production.  The reason is that people often test their component in dev, and then assume it will work in production.  If refs on stateless components worked in dev but not prod, then a user's code could break in production if they were relying on that ref.  This is why we don't provide refs in dev.  As a side note: Ideally unit tests are testing a unit's public API (not internals), so you shouldn't need to do anything in a unit test that couldn't/wouldn't happen in a production environment.  And you probably want your component tests to be in an environment that's as close to production as possible.  Personally, I like webdriver for this, since it's closer to prod, but scry works too.  But regardless, React components would/should/will NOT expose refs in dev if they aren't going to expose them in prod.

**Comment by joeybaker** [comment]
@oliverwoodings what you say about dev/prod parity for refs makes sense. What do you think about the proposal to strip out `testRef` for prod?

**Comment by oliverwoodings** [comment]
@jimfb thanks for getting involved in the discussion! Always nice to see the react team out in the community :) 

I actually agree with everything you say. I think maybe we have crossed wires a bit about where we are taking reacts-test-tree now and where we would like to see react heading. 

What the discussion in this thread has really boiled down to and revealed is:

* refs actually shouldn't be used for testing, or in fact anything except edge cases
* what we need is a testing system that behaves like refs, but does not affect the api available to a component (which is what we are actually implementing now in react-test-tree) 

To clarify my last point: right now, you can access refs from inside a component. What if you had a new type of property called testRef that was not accessible by a component (or its parents) and was only used internally by react for testing? This would mean the component itself behaves no differently in production vs dev vs testing, since you are literally just decorating elements so they can be tested easily.

**Comment by jimfb** [comment]
@oliverwoodings To be honest, I don't think I have enough context on this project to make an intelligent recommendation.  It sounds to me like you guys are asking good questions and working towards finding good solutions.

I'm not sure that I understand what `testRef` gets you above/beyond what a [callback ref](https://facebook.github.io/react/docs/more-about-refs.html#the-ref-callback-attribute) gets you, other than the ability to strip it in production.  (Stripping in production seems only marginally better than having callback refs call into your library, which could be replaced with a noop function in production).  Maybe I'm just missing something because I don't have any background on this project.

FWIW, it sounds like what you guys really want is an ability to (in dev) get references to all the instances in the render tree.  We've been talking about a new devtools API that would emit such information as the components render, thereby allowing you to build up your instance/ref tree during tests.  Unfortunately, such an API is not likely to happen before the end of the year.

**Comment by joeybaker** [comment]
@jimfb That API sounds good. Is there a good place to follow that discussion?

**Comment by jimfb** [comment]
@joeybaker There is a react github issue called "Umbrella: New Devtools API", but that's less of a discussion area and more of an area for tracking progress toward completion (much like our umbrella issues for a release).  Realistically, I still need to meet with Sebastian so we can pound out a reasonable proposal, and we'll be sure to post that info once we have something concrete for people to chew on.  Like I said, probably next half at the earliest.

**Comment by joeybaker** [comment]
Cool. Thanks @jimfb!

**Comment by oliverwoodings** [comment]
@jimfb the main reason for using `testRef` instead of `ref` is that `testRef` can be used on stateless components

**Comment by oliverwoodings** [comment]
That new devtools API does sound ideal, i'll keep an eye out for developments on it!

**Comment by goncalvesjoao** [comment]
@jimfb thank you so much for your input, I've tried using ref callbacks on stateless components but since stateless components don't have a "this" I can't do something like the following:

```javascript
const StatelessComponent = ({ onButton1, onButton2 }) => (
  <div>
    <button onClick={ onButton1 } />
    <button ref={(button2) => this.button2 = button2} onClick={ onButton2 } />
  </div>
);

// TypeError: Cannot set property 'button2' of undefined
```

I guess the simple question is how can I do something like the following example bellow, without my spec being couple to the components dom structure.

```javascript
import chai from 'chai';
import React from 'react';
import sinon from 'sinon';
import TestUtils from 'react-addons-test-utils';
import { findDOMNode } from 'react-dom';

const StatelessComponent = ({ onButton1, onButton2 }) => (
  <div>
    <button onClick={ onButton1 } />
    <button onClick={ onButton2 } />
  </div>
);

describe('StatelessComponent', () => {

  it('onButton1 and onButton2 props should be called according to their buttons', () => {
    const button1Spy = sinon.spy();
    const button2Spy = sinon.spy();

    const wrapper = TestUtils.renderIntoDocument(<div><StatelessComponent onButton1={ button1Spy } onButton2={ button2Spy } /></div>);

    const statelessComponent = findDOMNode(wrapper).children[0];
    const button1 = statelessComponent.children[0];
    const button2 = statelessComponent.children[1];

    TestUtils.Simulate.click(button1);

    chai.expect(button1Spy.called).to.be.true;
    chai.expect(button2Spy.called).to.be.false;
  });

});
```

If my fellow designers decide that button2 should be on top of button1
```javascript
const StatelessComponent = ({ onButton1, onButton2 }) => (
  <div>
    <button onClick={ onButton2 } />
    <button onClick={ onButton1 } />
  </div>
);
```
my specs will break although the component is still doing its job correctly.

I may be missing something but currently there's no solution for this right?

Thanks in advance to you @jimfb and all of you guys: @oliverwoodings , @aguestuser , @joeybaker , @activars

**Comment by jimfb** [comment]
@goncalvesjoao Right, but if I understand correctly, the issue is primarily that you are deciding which button to click based on order rather than button function.  That's not what a user would do, so that's not what your test should do.

```
const StatelessComponent = ({ onButton1, onButton2 }) => (
  <div>
    <button onClick={ onButton2 }>Button2</button>
    <button onClick={ onButton1 }>Button1</button>
  </div>
);
```

If you find the button based on the text within the button (ie. what the button "says", which is the same way that your users will find the buttons) then everything will work correctly (assuming I'm not miss-understanding your question).

**Comment by goncalvesjoao** [comment]
How can I find the button through its text, cycle through all children until one of them yields textContent === 'Button1' for instance?

Ok, that could work.
But the spec will still break if the designer team changes the copy of the button, the component logic is correct but the spec will break.

Something like css selector would be great like

```javascript
const StatelessComponent = ({ onButton1, onButton2 }) => (
  <div>
    <button id="button1" onClick={ onButton2 }>Button2</button>
    <button id="button2" onClick={ onButton1 }>Button1</button>
  </div>
);
```

const button1 = findDOMNode(statelessComponent, '#button1');

PS: Sorry to be replying in here, I guess the correct place to discuss this is at: https://github.com/facebook/react/issues/4972

**Comment by jimfb** [comment]
Yes, looping through, or using automated selectors as part of a test library/framework.  The test libraries we provide are primarily for our own use within the core.  If they happen to work for other people, great, but that was not the intent.  My guess is that you'll quickly outgrow the React testutils, and should try to find a real testing framework if you want to do automated testing.

But yes, UI tests are always fragile for exactly the reasons you've identified.

Personally, I like http://www.seleniumhq.org/ and `selenium webdriver`, which provide some utilities to make it easy to do things like select the desired button and simulate clicking on it.

**Comment by oliverwoodings** [comment]
Closed by #31
